rfcs/text/0019_lifecycle_preboot.md

```diff
@@ -0,0 +1,246 @@
+
+1. JS bundles routes will only include those exposed by the preboot plugins
+
+2. Default route for the preboot HTTP server will be hardcoded to the root path (`/`) since we cannot rely on the default value of the `defaultRoute` UI setting (`/app/home`)
+
+3. Main application route (`/app/{id}/{any*}`) will be replaced with the catch-all route (`/{path*}`). The reason is that if the user tries to access Kibana with a legit primary application URL (e.g. `/app/discover/?parameters`) while Kibana is still at the `preboot` stage, they will end up with `Application is not found` error. Instead, with the catch-all route, Kibana will capture the original URL in the `next` query string parameter and redirect the user to the root (e.g. `/?next=%2Fapp%2Fdiscover%2F%3Fparameters`). This will allow us to automatically redirect the user back to the original URL as soon as Kibana is ready. The main drawback and limitation of this approach are that there can be only one root-level preboot application. We can lift this limitation in the future if we have to though, for example, to support post-preboot Saved Objects migration UI or something similar.
+
+Serving a proper Kibana application on the root route of the preboot HTTP server implies that we'll also have a chance to replace the static `Kibana server is not ready yet` string with a more helpful and user-friendly application.
+
+### 3.2.9 Preboot service
+
+To support interactive applications at the `preboot` stage we should allow preboot plugins to pause Kibana startup sequence. This functionality will be exposed by the new Preboot service, and will be available to the preboot plugins via `CorePrebootSetup`. Preboot plugins will be able to provide a promise to hold a boot sequence for as much as needed, and also let Kibana know if it has to reload configuration before it enters the `setup` stage.
+
+```ts
+export interface PrebootServiceSetup {
+  readonly isBootOnHold: () => boolean;
+  readonly holdBootUntilResolved: (
+    promise: Promise<{ shouldReloadConfig: boolean } | undefined>
+  ) => void;
+}
+```
+
+Internal Preboot service contract will also expose `waitUntilCanBoot` method that bootstrap process can use to know when it can proceed to `setup` and `start` stages:
+
+```ts
+// NOTE(azasypkin): some existing properties have been omitted for brevity.
+export interface InternalPrebootServiceSetup {
+  readonly waitUntilCanBoot: () => Promise<{ shouldReloadConfig: boolean }>;
+}
+```
+
+### 3.2.10 Bootstrap
+
+We'll update Kibana bootstrap sequence to include `preboot` stage and to conditionally reload configuration before proceeding to `setup` and `start` stages:
+
+```ts
+// NOTE(azasypkin): some functionality and checks have been omitted for brevity.
+const { preboot } = await root.preboot();
+
+const { shouldReloadConfig } = await preboot.waitUntilCanBoot();
+if (shouldReloadConfig) {
+  await reloadConfiguration('pre-boot request');
```

Comment:
> Could we get into a scenario where the written config was valid preboot time, but then something within the environment causes the config to become invalid?

We can, but I see only two reasons when this can happen:

* Preboot plugins produce invalid config due to a bug
* User changes config outside of Kibana for some reason, while Kibana is at the `preboot` stage

These look like unrecoverable issues to me, and I'd prefer Kibana to crash as it does today when config validation fails. For example, if HTTP config fails validation we won't be able to even start preboot HTTP server.

> Having preboot run again would give us the opportunity to correct a misconfigiured instance, or at least show a status page.

That would be a great feature for Kibana for sure, but I feel like it's a bit out of scope for now. The near future plan is to make very few config changes and every change should be extensively validated.

Having said that, you're raising a very important point that seems to be outside of this RFC, but we'll need to take care of it as soon as we start working on the `setup` plugin: preboot plugins should have access to schemas (or validation methods) for the Core configuration they may change. Same goes for security configuration (encryption keys, auth providers etc.)